internal/kgateway/extensions2/plugins/routepolicy/route_policy_plugin.go

```
@@ -247,14 +249,44 @@ func (p *routePolicyPluginGwPass) ApplyForRoute(ctx context.Context, pCtx *ir.Ro
 				},
 			},
 		}
-		outputRoute.GetTypedPerFilterConfig()[metadataRouteTransformation], _ = utils.MessageToAny(setmetaTransform)
+		pCtx.TypedFilterConfig.AddTypedConfig(metadataRouteTransformation, setmetaTransform)
 
 		p.setTransformationInChain = true
 	}
 
-	// TODO: err/warn/ignore if targetRef is set on non-AI Backend
+	if policy.spec.AI != nil {
+		var aiBackends []*v1alpha1.Backend
+		// check if the backends selected by targetRef are all AI backends before applying the policy
+		for _, backend := range pCtx.In.Backends {
+			if backend.Backend.BackendObject == nil {
+				// could be nil if not found or no ref grant
+				continue
+			}
+			b, ok := backend.Backend.BackendObject.Obj.(*v1alpha1.Backend)
+			if !ok {
+				// AI policy cannot apply to kubernetes services
+				errs = append(errs, fmt.Errorf("targetRef cannot apply to %s backend. AI RoutePolicy must apply only to AI backend", backend.Backend.BackendObject.GetName()))
+				continue
```

do we really need to handle this? can we just ensure that e.g. adding an AI RoutePolicy to a non-AI Backend is a no-op ??
cc @yuval-k

---

I think it's slightly different from other RoutePolicies (transformations, etc.) that can be applied regardless of what the backend type is? But we could make this a no-op (not log anything, assume the extproc config is already there, skip if it's not)

---

IMO, having validation logic and logging that it's not supported is a better UX than silently ignoring it and the user wonders why their config isn't working.

---

Discussed offline- if we return an error the route will be dropped, which isn't the behavior we want. Will log this as a warning and add a TODO to add this as a status warning on the resource.

---

>IMO, having validation logic and logging that it's not supported is a better UX than silently ignoring it and the user wonders why their config isn't working. At least in this case, the user can fix the config and silence the error log msg.

@timflannagan are you in favor of the original behavior of dropping the route entirely?

---

Discussed offline, will revert to returning error here for now. Eventually will do route replacement to return 500 with error message.